File: polyglot/DesignPatterns/AbstractFactory/AbstractFactory.scala
// Abstract Factory design pattern
// the Abstract Factory pattern looks like the factory pattern 
// applied to the production of various factories

abstract class Shape {
  def draw()
}

abstract class AbstractShape(val color: AbstractShape.Color.Color) extends Shape {
}

object AbstractShape {
  // http://www.scala-lang.org/api/current/scala/Enumeration.html
  object Color extends Enumeration {
    type Color = Value            // needed to 'Color' can be used as a type
    val RED, GREEN, BLUE = Value  // '= Value' ??? what is that?  
  }
  
  import Color._
  def asString(color: Color): String = {
    color match {
      case RED    => "red"
      case GREEN  => "green"
      case BLUE   => "blue"
      case _      => "unknown color"
    }  
  }
}

class Rectangle(override val color: AbstractShape.Color.Color) 
  extends AbstractShape (color) {
  import AbstractShape._
  def draw() = {
    println("Drawing %s rectangle" format asString(color))
  }
}

class Square(override val color: AbstractShape.Color.Color) 
  extends AbstractShape (color) {
  import AbstractShape._
  def draw() = {
    println("Drawing %s square" format asString(color))
  }
}

class Circle(override val color: AbstractShape.Color.Color) 
  extends AbstractShape (color) {
  import AbstractShape._
  def draw() = {
    println("Drawing %s circle" format asString(color))
  }
}

// using the template method pattern here, as the actual
// behaviour of 'getShape' will be defined via specialization
// of virtual method getColor through subclassing
abstract class AbstractShapeFactory {
  def getColor(): AbstractShape.Color.Color
  def getShape(shapeType: String): Shape = {
    if(shapeType == null) return null

    if(shapeType.equalsIgnoreCase("CIRCLE")){
      return new Circle(getColor())
    } else if(shapeType.equalsIgnoreCase("RECTANGLE")){
      return new Rectangle(getColor())
    } else if(shapeType.equalsIgnoreCase("SQUARE")){
      return new Square(getColor())
    }
    return null
  }
}

// However the benefit of subclassing over maintaining
// 'mColor' state in base class is not that clear in this simple case
// It is propably beneficial when the distinction between various
// families of widgets (Shape) goes well beyond color difference.
class RedShapeFactory extends AbstractShapeFactory {
  def getColor() = AbstractShape.Color.RED 
}

class GreenShapeFactory extends AbstractShapeFactory {
  def getColor() = AbstractShape.Color.GREEN 
}

class BlueShapeFactory extends AbstractShapeFactory {
  def getColor() = AbstractShape.Color.BLUE 
}

// Factory of factories. The Abstract Factory design pattern is a case
// of Factory design pattern applied to various factory types.
class FactoryProducer {
  def getFactory(factoryType: String): AbstractShapeFactory = {
    if(factoryType == null) return null
    if(factoryType.equalsIgnoreCase("RED")){
      return new RedShapeFactory
    } else if(factoryType.equalsIgnoreCase("GREEN")){
      return new GreenShapeFactory
    } else if(factoryType.equalsIgnoreCase("BLUE")){
      return new BlueShapeFactory
    }
    return null
  }
}

object AbstractFactory {
  import AbstractShape.Color._
 
  def main(args: Array[String]) = {
    val producer = new FactoryProducer
    // producing set of red widgets
    val redFactory = producer.getFactory("Red")
    val shape1 = redFactory.getShape("CIRCLE")
    val shape2 = redFactory.getShape("RECTANGLE")
    val shape3 = redFactory.getShape("SQUARE")
    shape1.draw()
    shape2.draw()
    shape3.draw()

    // producing set of green widgets
    val greenFactory = producer.getFactory("Green")
    val shape4 = greenFactory.getShape("CIRCLE")
    val shape5 = greenFactory.getShape("RECTANGLE")
    val shape6 = greenFactory.getShape("SQUARE")
    shape4.draw()
    shape5.draw()
    shape6.draw()
 
    // producing set of blue widgets
    val blueFactory = producer.getFactory("Blue")
    val shape7 = blueFactory.getShape("CIRCLE")
    val shape8 = blueFactory.getShape("RECTANGLE")
    val shape9 = blueFactory.getShape("SQUARE")
    shape7.draw()
    shape8.draw()
    shape9.draw()
  }
}

File: poly/DesignPatterns/AbstractFactory/abstractFactory.clj
; Abstract Factory design pattern
(ns abstractFactory (:gen-class))


(defn Shape [hook]
  (letfn
    ; interface
    [(this [m]
       (cond (= m :draw) (hook)
             :else (println "Shape: unknown operation error")))]
    ; returning interface
    this))

(defn AbstractShape [hook color]
  (let
    ; data
    [_base (Shape hook)
     _color color]
    (letfn
      ;interface
      [(this [m]
         (cond (= m :color) _color
               (= m :asString) asString
               :else (_base m)))
       (asString [color]
         (cond (= color :RED) "red"
               (= color :GREEN) "green"
               (= color :BLUE) "blue"
               :else "unknown color"))]
      ; returning interface
      this)))

(def Rectangle      ; constructor
  (letfn
    ; object created from data is message passing interface
    [(this [data]
       (fn [m] 
         (cond (= m :draw) (draw data)
         :else ((base data) m)))) ; passing message to base object
     (base [data] data)           ; data = base object
     (draw [data]
       (print "Drawing ")
       (print (((base data) :asString) ((base data) :color)))
       (println " rectangle"))]
    ; returning single constructor
    (fn [color] (this (AbstractShape (fn [] :ignored) color)))))

(def Square      ; constructor
  (letfn
    ; object created from data is message passing interface
    [(this [data]
       (fn [m] 
         (cond (= m :draw) (draw data)
         :else ((base data) m)))) ; passing message to base object
     (base [data] data)           ; data = base object
     (draw [data]
       (print "Drawing ")
       (print (((base data) :asString) ((base data) :color)))
       (println " square"))]
    ; returning single constructor
    (fn [color] (this (AbstractShape (fn [] :ignored) color)))))

(def Circle      ; constructor
  (letfn
    ; object created from data is message passing interface
    [(this [data]
       (fn [m] 
         (cond (= m :draw) (draw data)
         :else ((base data) m)))) ; passing message to base object
     (base [data] data)           ; data = base object
     (draw [data]
       (print "Drawing ")
       (print (((base data) :asString) ((base data) :color)))
       (println " circle"))]
    ; returning single constructor
    (fn [color] (this (AbstractShape (fn [] :ignored) color)))))


(defn AbstractShapeFactory [hook]
  (let [_hook hook]
    (letfn
      [(this [m]
         (cond (= m :getColor) (_hook)
               (= m :getShape) getShape
               :else (println "AbstractShapeFactory: unknown operation error")))
       (getShape [shapeType]
         (cond (= (.toUpperCase shapeType) "CIRCLE") (Circle (_hook))
               (= (.toUpperCase shapeType) "RECTANGLE")(Rectangle(_hook))
               (= (.toUpperCase shapeType) "SQUARE") (Square (_hook))
               :else nil))]
      ; returning interface
      this)))

(defn RedShapeFactory []
  (AbstractShapeFactory (fn [] :RED)))

(defn GreenShapeFactory []
  (AbstractShapeFactory (fn [] :GREEN)))

(defn BlueShapeFactory []
  (AbstractShapeFactory (fn [] :BLUE)))

(defn FactoryProducer []
  (letfn
    [(this [m]
     (cond (= :getFactory) getFactory
           :else (println "FactoryProducer: unknown operation error")))
     (getFactory [factoryType]
       (cond (= (.toUpperCase factoryType) "RED") (RedShapeFactory)
             (= (.toUpperCase factoryType) "GREEN") (GreenShapeFactory)
             (= (.toUpperCase factoryType) "BLUE") (BlueShapeFactory)
             :else nil))]
    ; returning interface
    this))


(defn -main []

(def producer (FactoryProducer))

; producing set of red widgets
(def redFactory ((producer :getFactory) "Red"))
(def shape1 ((redFactory :getShape) "CIRCLE"))
(def shape2 ((redFactory :getShape) "RECTANGLE"))
(def shape3 ((redFactory :getShape) "SQUARE"))
(shape1 :draw)
(shape2 :draw)
(shape3 :draw)

; producing set of green widgets
(def greenFactory ((producer :getFactory) "Green"))
(def shape4 ((greenFactory :getShape) "CIRCLE"))
(def shape5 ((greenFactory :getShape) "RECTANGLE"))
(def shape6 ((greenFactory :getShape) "SQUARE"))
(shape4 :draw)
(shape5 :draw)
(shape6 :draw)

; producing set of blue widgets
(def blueFactory ((producer :getFactory) "Blue"))
(def shape7 ((blueFactory :getShape) "CIRCLE"))
(def shape8 ((blueFactory :getShape) "RECTANGLE"))
(def shape9 ((blueFactory :getShape) "SQUARE"))
(shape7 :draw)
(shape8 :draw)
(shape9 :draw))

;(-main)



